### PR TITLE
Name: add support for multi-value RDNs

### PR DESCRIPTION
**Pull request by frasertweedale**
Implement initial design proposal for #3199.

### CODE REVIEW

**Comment by alex** [comment]
Jenkins, ok to test.

**Comment by frasertweedale** [comment]
X509_NAME_ENTRY * became an opaque type in openssl 1.1.0;
there a function to get at the needed value but I just need to figure out
how to handle the different versions

**Comment by frasertweedale** [comment]
phew, green at last

File: src/cryptography/hazmat/backends/openssl/decode_asn1.py [hunk 1]
```diff
@@ -548,17 +556,17 @@ def _decode_crl_distribution_points(backend, cdps):
             else:
                 rns = cdp.distpoint.name.relativename
                 rnum = backend._lib.sk_X509_NAME_ENTRY_num(rns)
-                attributes = []
+                attributes = set()
```
Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/decode_asn1.py]: Could you clarify what's going on here? As I understand it this change will make every name entry be added to a single set, which means that all CRLDistributionPoint relativename names are multi-valued?

File: src/_cffi_src/openssl/x509name.py [hunk 2]
```diff
@@ -76,4 +77,19 @@
 """
 
 CUSTOMIZATIONS = """
+#if CRYPTOGRAPHY_OPENSSL_110_OR_GREATER && !defined(LIBRESSL_VERSION_NUMBER)
+int Cryptography_X509_NAME_ENTRY_set(X509_NAME_ENTRY *ne) {
+    return X509_NAME_ENTRY_set(ne);
+}
+#else
+struct x509_name_entry_st {
```
Comment (frasertweedale) [inline src/_cffi_src/openssl/x509name.py]: I think you're right.  Removed it.

File: src/cryptography/hazmat/backends/openssl/decode_asn1.py (hunk 1 above)
Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/decode_asn1.py]: Sorry about the delay. I am okay with making a breaking change here if we implemented it incorrectly (which it appears we, or more specifically, I, did). I also agree that should be a separate PR. To make sure I understand this then, you have it all adding to a single set so that it behaves as much like an RDN as possible, correct?

File: docs/x509/reference.rst [hunk 3]
```diff
@@ -1112,6 +1124,12 @@ X.509 CSR (Certificate Signing Request) Builder Object
         <NameAttribute(oid=<ObjectIdentifier(oid=2.5.4.10, name=organizationName)>, value=u'Test Certificates 2011')>
         <NameAttribute(oid=<ObjectIdentifier(oid=2.5.4.3, name=commonName)>, value=u'Good CA')>
 
+    .. attribute:: rdns
```
Comment (reaperhulk) [inline docs/x509/reference.rst]: Now that `Name` is an iterable of `RelativeDistinguishedName` objects does it make sense to keep this attribute?

File: src/_cffi_src/openssl/x509name.py [hunk 4]
```diff
@@ -17,7 +17,7 @@
 TYPES = """
 typedef ... Cryptography_STACK_OF_X509_NAME_ENTRY;
 typedef ... X509_NAME;
-typedef ... X509_NAME_ENTRY;
+typedef struct X509_name_entry_st X509_NAME_ENTRY;
```
Comment (reaperhulk) [inline src/_cffi_src/openssl/x509name.py]: This can stay opaque.

File: docs/x509/reference.rst (hunk 3 above)
Comment (frasertweedale) [inline docs/x509/reference.rst]: Iterating `RelativeDistinguishedName` still yields `NameAttribute` values, for compatibility (and it still adequately covers the 99%** use case where each RDN has exactly one attribute).

So ``rdns`` lets you get the sequence of actual RDNs, if you need them.

** guessing :)

File: src/_cffi_src/openssl/x509name.py (hunk 4 above)
Comment (frasertweedale) [inline src/_cffi_src/openssl/x509name.py]: typedef opacified :)

File: docs/x509/reference.rst [hunk 5]
```diff
@@ -1112,6 +1124,12 @@ X.509 CSR (Certificate Signing Request) Builder Object
...
+
+        .. versionadded:: 1.6
+
+        :type: :class:`RelativeDistinguishedName`
```
Comment (reaperhulk) [inline docs/x509/reference.rst]: This returns an iterable of RDNs right?

Comment (frasertweedale) [inline docs/x509/reference.rst]: Yep; thanks. PR updated.

**Comment by reaperhulk** [comment]
Thanks again for your work on this!
